test/ecto/embedded_test.exs

```diff
@@ -31,10 +44,142 @@ defmodule Ecto.EmbeddedTest do
+    profile = changeset.changes.profile
+    assert profile.changes == %{name: "new", id: "new"}
+    assert profile.errors  == []
+    assert profile.status  == :insert
+    assert profile.valid?
+    assert changeset.valid?
+  end
+
+  test "cast embeds_one with custom changeset" do
+    changeset = cast(%Author{}, %{"profile" => %{"name" => "michal"}},
+                     [profile: :optional_changeset])
+    profile = changeset.changes.profile
+    assert profile.changes == %{name: "michal"}
+    assert profile.errors  == []
+    assert profile.status  == :insert
+    assert profile.valid?
+    assert changeset.valid?
+
+    changeset = cast(%Author{}, %{"profile" => %{}}, [profile: :optional_changeset])
+    profile = changeset.changes.profile
+    assert profile.changes == %{}
+    assert profile.errors  == []
+    assert profile.status  == :insert
+    assert profile.valid?
+    assert changeset.valid?
+  end
+
+  test "cast embeds_many with only new models" do
+    changeset = cast(%Author{}, %{"profiles" => [%{"name" => "michal"}]}, ~w(profiles))
+    [profile_change] = changeset.changes.profiles
+    assert profile_change.changes == %{name: "michal"}
+    assert profile_change.errors  == []
+    assert profile_change.status  == :insert
+    assert profile_change.valid?
+    assert changeset.valid?
+  end
+
+  test "cast embeds_many with custom changeset" do
+    changeset = cast(%Author{}, %{"profiles" => [%{"name" => "michal"}]},
+                     [profiles: :optional_changeset])
+    [profile_change] = changeset.changes.profiles
+    assert profile_change.changes == %{name: "michal"}
+    assert profile_change.errors  == []
+    assert profile_change.status  == :insert
+    assert profile_change.valid?
+    assert changeset.valid?
+  end
+
+  # Please note the order is important in this test.
+  test "cast embeds_many changing models" do
+    profiles = [%Profile{name: "michal", id: "michal"},
+                %Profile{name: "unknown", id: "unknown"},
+                %Profile{name: "other", id: "other"}]
+    params = [%{"id" => "new", "name" => "new"},
+              %{"id" => "unknown", "name" => nil},
+              %{"id" => "other", "name" => "new name"}]
+
+    changeset = cast(%Author{profiles: profiles}, %{"profiles" => params}, ~w(profiles))
+    [new, unknown, other, michal] = changeset.changes.profiles
+    assert new.changes == %{name: "new"}
+    assert new.status == :insert
+    assert new.valid?
+    assert unknown.model.id == "unknown"
+    assert unknown.errors == [name: "can't be blank"]
+    assert unknown.status == :update
+    refute unknown.valid?
+    assert other.model.id == "other"
+    assert other.status == :update
+    assert other.valid?
+    assert michal.model.id == "michal"
+    assert michal.status == :delete
+    assert michal.valid?
+    refute changeset.valid?
+  end
+
+  test "cast embeds_many with invalid params" do
```

Comment:
Let's also add an assertion when the parameter is nil. We shuold mark it as invalid too.